Enable PSA tests on K64F/K66F and fix missing PSA Crypto init in TLSSocketWrapper

LDong-Arm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

This PR includes two main changes:

* Enable PSA tests on Mbed OS PSA targets (K64F and K66F):

  When using the `mbed test` command to build and run tests, some targets have additional configurations/overrides defined in `tools/test_configs/`:
    * `target_configs.json` lists which targets support which configs.
    * `config_paths.json` maps the name of each config to the JSON file to
use.

  By default, only `default_test_configuration` from `target_configs.json` gets used (i.e. automatically picked up by Mbed CLI 1) when building and running tests. Others listed in `test_configuration` need to be switched via `--test-config <NAME>`.

  Enable Experimental API in the default configurations on K64F and K66F in order to test Mbed OS PSA. Any existing configs are kept, which is why `HeapBlockDeviceAndEthernetAndExperimental.json` is created for K64F.

* Fix missing PSA Crypto initialization in `TLSSocketWrapper`:

  With Experimental API/PSA enabled as above, `TLSSocket` does not work anymore and its test `connectivity/netsocket/tests/TESTS/netsocket/tls` fails. This is because Mbed TLS uses PSA Crypto for cryptographic operations when available, but PSA Crypto needs to be initialized first (and it currently isn't).

  To fix this, we call `psa_crypto_init()` in `TLSSocketWrapper` when `MBEDTLS_USE_PSA_CRYPTO`.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

None

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR
    
Manual testing of CMake support (not yet automated by CTest or run in CI):
Build:
```
cd connectivity/netsocket/tests/TESTS/netsocket/tls
mbedtools configure -t GCC_ARM -m K64F --mbed-os-path <path-to>/mbed-os --app-config <path-to>/mbed-os/tools/test_configs/experimental.json
cmake --build .
```
Run (K64F with Ethernet):
```
mbedhtrun -f mbed-connectivity-netsocket-tls.bin -p <port> -d <mount>
```
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

@ARMmbed/mbed-os-core @Patater @saheerb 

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

LDong-Arm committed:
target_configs.json: Remove unreachable nsapi entries

In `tools/test_configs/`, target-specific test configurations are
defined in `target_configs.json` and parsed by `__init__.py`. The
latter only makes use of `default_test_configuration` (default test
configuration to use) and `test_configurations` (more configurations
selectable via `mbed test --test-config <config>`. Anything else
is ignored, including nsapi, so this commit cleans up dead entries.

LDong-Arm committed:
test configs: Test Experimental API on K64F/K66F

When using the `mbed test` command to build and run tests, some
targets have additional configurations/overrides defined in
`tools/test_configs/`:
* `target_configs.json` lists which targets support which configs.
* `config_paths.json` maps the name of each config to the JSON file to
use.

By default, only `default_test_configuration` from
`target_configs.json` gets used when building and running tests.
Others listed in `test_configuration` need to be switched via
`--test-config <NAME>`.

This commit enables Experimental API in the default configurations of
K64F and K66F in order to test Mbed OS PSA. Any existing configs are
kept, which is why `HeapBlockDeviceAndEthernetAndExperimental.json` is
created for K64F.

LDong-Arm committed:
platform_mbed.h: Fix and align EXPERIMENTAL_API check

Some of the lines in `platform_mbed.h` only have `FEATURE_PSA`
checked, which is always set for Mbed OS PSA targets but the PSA
APIs are not actually available unless `FEATURE_EXPERIMENTAL_API`
is also enabled. To fix this and improve readability, group all
PSA-related lines and check both macros.

LDong-Arm commented on connectivity/mbedtls/platform/inc/platform_mbed.h:
```diff
@@ -58,6 +49,17 @@
 
 #define MBEDTLS_ENTROPY_HARDWARE_ALT
+
+#endif  // defined(FEATURE_EXPERIMENTAL_API) && defined(FEATURE_PSA)
```
The diff rendered by git isn't helpful here. I moved the PSA-related lines above instead of the TRNG lines below, but git decides to show it the other way round... Please check the whole file during review.

(I'm talking about the commit "platform_mbed.h: Fix and align EXPERIMENTAL_API check")

Patater commented on connectivity/mbedtls/platform/inc/platform_mbed.h:
```diff
@@ -50,6 +50,8 @@
 
 #define MBEDTLS_ENTROPY_HARDWARE_ALT
 
+#define MBEDTLS_PLATFORM_SETUP_TEARDOWN_ALT
```
This is enabled in "mbedtls: Initialize PSA Crypto if available", even for non-PSA configurations, but no implementation is provided for non-PSA (and non-cryptocell) configurations.

LDong-Arm replied:
This is enabled for PSA only - see lines 24 and 55. This line is within the `#ifdef` - `#endif` for PSA.

Patater replied:
OK, I missed this was in a PSA only section.

Patater commented on platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/src/mbedtls_crypto_setup.c:
```diff
@@ -0,0 +1,35 @@
+
+// Required by mbedtls_platform_setup()
+
+int crypto_platform_setup(crypto_platform_ctx *unused)
```
Should we have an ifdef to catch if cryptocell is enabled to provide a more friendly error message than duplicate symbol definition? We can't have both PSA and cryptocell enabled, unless we chain together their initializers.

LDong-Arm replied:
Good point, I need to check whether Mbed TLS can use CryptoCell and PSA Crypto at the same time. If it can, we need to make sure they are both initialised.

LDong-Arm replied:
It's tricky to initialise both CryptoCell and PSA Crypto here, because any crypto hardware (e.g. CryptoCell) relies on `MBEDTLS_PLATFORM_SETUP_TEARDOWN_ALT` (from Mbed TLS) to add initialisation hooks, but the same macro is for overriding `mbedtls_platform_setup()` for PSA Crypto too. So, there's no way to check if only PSA is available or PSA and some crypto hardware both need to be initialised - we just don't have any macro or anything for this.

Introducing a new config/macro would be a breaking change - it's easy to change CryptoCell as it's in Mbed OS, but if a third-party project has its own crypto hardware integrated with the same hook, it will break. Breaking changes (without a backward compatibility mechanism) are only allowed in Mbed OS between major releases (e.g. 5 -> 6).

So I just've completely changed the approach, and simply call `psa_crypto_init()` in `TLSSocketWrapper` if `MBEDTLS_USE_PSA_CRYPTO`. This is what [an example program of Mbed TLS](https://github.com/ARMmbed/mbedtls/blob/8682faeb092f32020602514fb0374d2ad1ce541a/programs/ssl/ssl_client2.c#L739-L748) does too.

LDong-Arm committed:
TLSSocketWrapper: Initialize PSA Crypto if used by Mbed TLS

When `MBEDTLS_USE_PSA_CRYPTO` is set, Mbed TLS uses the PSA Crypto API
where possible. It is necessary to initialize PSA Crypto beforehand.

Patater commented on connectivity/netsocket/source/TLSSocketWrapper.cpp:
```diff
@@ -41,6 +45,16 @@ TLSSocketWrapper::TLSSocketWrapper(Socket *transport, const char *hostname, cont
+    // function.
+    psa_status_t status = psa_crypto_init();
+    if (status != PSA_SUCCESS) {
+        tr_err("psa_crypto_init() failed (" PRIu32 ")", status);
```
Should we use `print_mbedtls_error()` here like we do in other functions?

LDong-Arm replied:
PSA status is different from Mbed TLS error, and `print_mbedtls_error()` uses `mbedtls_strerror()` to get error strings which doesn't support translating PSA errors. ([Those modules](https://github.com/ARMmbed/mbedtls/blob/v2.26.0/scripts/generate_errors.pl#L47-L54) are supported.)

Patater replied:
OK, too bad it doesn't support PSA errors.
